src/utils/recompiler.ts

```diff
@@ -0,0 +1,201 @@
+): Promise<Set<string>> => {
+    // Set to save contracts with changed content & unavailable artifacts
+    const changed: Set<string> = new Set();
+    const { starknetSources: defaultSourcesPath } = hre.config.paths;
+    const artifactsDir = getArtifactPath(defaultSourcesPath, hre);
+    // traverse on artifacts directory
+    const artifactsList = await traverseFiles(artifactsDir, "*.json");
+    for (const name in newCacheEntry) {
+        const outputPath = newCacheEntry[name].outputPath;
+        const abiPath = newCacheEntry[name].abiPath;
+        if (!artifactsList.includes(outputPath) || !artifactsList.includes(abiPath)) {
+            changed.add(name);
+        }
+    }
+
+    return changed;
+};
+
+// Compile changed contracts
+const compileChangedContracts = async (
+    args: TaskArguments,
+    hre: HardhatRuntimeEnvironment,
+    newCacheEntry: CacheEntry,
+    oldCacheEntry: CacheEntry,
+    changed: Set<string>
+): Promise<void> => {
+    for (const contractName in newCacheEntry) {
+        // Add new contracts that are not in cache before
+        if (!oldCacheEntry[contractName]) {
+            changed.add(contractName);
+        }
+
+        // Add contracts that contiain a change in content
+        if (newCacheEntry[contractName].contentHash !== oldCacheEntry[contractName]?.contentHash) {
+            changed.add(contractName);
+        }
+    }
+
+    if (changed.size > 0) {
+        // Compiles contracts
+        console.log("Compiling contracts...");
+        const copyPaths = args.paths;
+        args.paths = [...changed];
+        await starknetCompileAction(args, hre);
+        // Restore paths
+        args.paths = copyPaths;
+    }
+};
+
+// Handles cache on Starknet cli calls
+export const handleCache = async (args: TaskArguments, hre: HardhatRuntimeEnvironment): Promise<boolean> => {
+    // If recompile is not enabled, do nothing
+    if (!hre.userConfig?.starknet?.recompile) return true;
+
+    // Get cache directory from hre.config
+    const { cache: cacheDirName } = hre.config.paths;
+
+    try {
+        const oldCacheEntry = await upsertFile(cacheDirName);
+        const newCacheEntry = await getContractHash(hre);
+        const changedContracts = await checkArtifacts(hre, newCacheEntry);
+        await compileChangedContracts(args, hre, newCacheEntry, oldCacheEntry, changedContracts);
+
+        return true;
+    } catch (error) {
+        // If there is an error, do not recompile
+        console.error(error);
+        return false;
+    }
+};
+
+// Save cache entry to file
+export const saveCacheEntry = async (
+    file: string,
+    output: string,
+    abi: string,
+    hre: HardhatRuntimeEnvironment
+): Promise<void> => {
+    const { cache: cacheDirName } = hre.config.paths;
+    const oldCacheEntry = await upsertFile(cacheDirName);
+    const newCacheEntry = await getHashEntry(file, output, abi);
+    const updatedCacheEntry = getUpdatedCashEntry(oldCacheEntry, newCacheEntry);
```

Comment:
Why not just save the new cache entry? Doesn't it contain all the necessary data for future recompilations? Also, the way the updating is done, it would never delete unused parts of the cache (of old, deleted source files).

Reply:
https://github.com/Shard-Labs/starknet-hardhat-plugin/pull/106/commits/61024201aafb0d528027251c3c6276a6afd9533e